### ISSUE TITLE
[feat] no-leaked-conditional-rendering: support not allowing boolean variant

### ISSUE THREAD

**Issue by czarandy** [Rel1cx/eslint-react#1840]
### Problem Description

Currently no-leaked-conditional-rendering hardcoded boolean in its allowedVariants list. However I would like to make it not allowed to have a boolean condition. 

The reason is that sometimes components have logic like:
```
if (children != null) {
 // do something
}
```

If we then pass in `isFoo && <Bar />` which resolves to `false`, then the component `children != null` check will pass, which may result in incorrect rendering.

### Alternative Solutions

There no way to use this rule without this option to do the boolean check. Currently I have just forked it locally.

### Feature Details

Feature: support not allowing boolean variant
I prefer to require ternaries for such cases to avoid these sorts of bugs. So it would be great if we could have an option to support this behavior, it seems easy enough. It could either be a more flexible option to ban arbitrary variants, or scoped to just boolean.

### Examples

It would ban uses like this:
```
<Foo>{isBoo && <Bar />}</Foo>
```

to use it you would just pass the option to the rule, e.g. `'no-boolean-variant'` or something would be a reasonable name.

### Evaluation Checklist

- [x] I have had problems that this feature would solve
- [x] I could not find a way to solve the problem with existing features or workarounds
- [x] I have thought very hard about potential edge cases and downsides, and they are acceptable
- [x] I think the feature is well-defined and would provide clear value to users

**Comment by Rel1cx** [comment]
The fundamental misconception here is translating a **"parent component's type requirement on child return values"** into a **"lint rule that bans boolean conditional rendering for everyone."**

That is the wrong direction.

`{isFoo && <Bar />}` is a [valid and widely used React pattern](https://react.dev/learn/conditional-rendering#logical-and-operator). The real problem is not the `&&` syntax itself. It is that **in certain specific scenarios, a parent component expects its child to never return a boolean**. That expectation is a component contract, and contracts should be enforced by the type system, not by a blanket lint rule.

### Why this falls short

Adding a rule option to ban the boolean variant would punish all users for one component's internal implementation detail. It would also block perfectly valid conditional rendering like `isLoading && <Spinner />`.

### What to do instead

If your parent relies on `children != null` checks and you must ensure a child never returns a boolean, remove `boolean` from that specific child's return type. Let TypeScript catch the mismatch at compile time.

```tsx
function ChildFoo(): Exclude<ReactNode, boolean> {
  return false && <div></div>; // Type error: boolean is not assignable to Exclude<ReactNode, boolean>
}

function ChildBar() {
  return false && <div></div>;
}

// Enforce at the component level
ChildBar satisfies () => Exclude<ReactNode, boolean>; // Type error

function ChildBaz() {
  return false && <div></div>;
}

function App() {
  // Or enforce at the usage site
  ChildBaz satisfies () => Exclude<ReactNode, boolean>; // Type error
  return (
    <ExpectChildNoBolean>
      <ChildBaz />
    </ExpectChildNoBolean>
  );
}
```

### Summary

| | Changing the lint rule | Constraining the return type |
|---|---|---|
| Scope | Bans `&&` for everyone | Targets only the specific child component |
| When it catches issues | Lint time | Compile time |
| Impact on valid code | Breaks legitimate patterns | No side effects |
| Intent | Hidden behind option names | Explicit in the type signature |

This feature request inverts the cost of the solution. **Use the type system to enforce component contracts.** Do not use a lint rule to prohibit a legitimate React pattern.

**Comment by czarandy** [comment]
In your own contributing guide you say "Don't use AI to respond to review comments (except for translations)." yet your comment is obviously AI generated. 

I could reply in detail but what's the point if you're just going to have AI respond anyway. Obviously it doesn't realize that not all components are controlled by me, so if I have a library I am using that doesn't handle `false` well I need to guard against it by preventing my own code from passing it in.

**Comment by Rel1cx** [comment]
> In your own contributing guide you say "Don't use AI to respond to review comments (except for translations)." yet your comment is obviously AI generated.
> 
> I could reply in detail but what's the point if you're just going to have AI respond anyway. Obviously it doesn't realize that not all components are controlled by me, so if I have a library I am using that doesn't handle `false` well I need to guard against it by preventing my own code from passing it in.

You're absolutely right!

**Comment by Rel1cx** [comment]
```tsx
function ThirdPartyLibrary({ children: React.ReactNode }) {
  // The children is a `type React.ReactNode = string | number | bigint | boolean | React.ReactElement<unknown, string | React.JSXElementConstructor<any>> | Iterable<React.ReactNode> | React.ReactPortal | Promise<AwaitedReactNode> | null | undefined` 
  // What are you expecting `if (children != null)` for?
  if (children != null) {
    // do something
  }
}
```

If it's a library not controlled by you, why not submit an issue to the library maintainer and make a local wrapper until the issue is resolved?

```tsx
// Only import it from here
function ExpectChildrenNoBoolean({ children: Exclude<ReactNode, boolean> }) {
  return <ThirdPartyLibrary>{children}</ThirdPartyLibrary>
}

function App() {
  return (
    <ExpectChildrenNoBoolean>
      {isFoo && <Bar />}
    </ExpectChildrenNoBoolean>
  );
}
```

Or just switch libraries:

```tsx
function ThirdPartyLibrary({ children: React.ReactNode }) {
  // The children is a `type React.ReactNode = string | number | bigint | boolean | React.ReactElement<unknown, string | React.JSXElementConstructor<any>> | Iterable<React.ReactNode> | React.ReactPortal | Promise<AwaitedReactNode> | null | undefined` 
  // What are you expecting `if (children != null)` for?
  if (children != null) {
    // do something
  }
}
```

This is about narrowing down the type from a union, not the complex issues that require you to go through the entire process described in https://www.github.com/acdlite/react/commit/204d3dd703f2ef21953e1509302456f3800a07db, https://www.github.com/reactwg/react-18/discussions/72, https://www.github.com/facebook/react/pull/21869, https://www.github.com/reactwg/react-18/discussions/75, or that the current type system cannot handle, requiring you to use Flow's Render Types–like solution shown at https://github.com/react-analyzer/react-annotation#render-types.

If a library lacks knowledge explicitly stated in the React docs, why would you choose it? If you still insist on using it, you've realized that you don't control the library, and similarly, you don't control this repo either. By properly submitting issues to the library where the root cause of the problem lies, you can help the library author grow and fill in the missing necessary knowledge, as well as help all users of the library. But you only want this problem not to trigger in your own code. Why be so self-centered?

Additionally, if you truly believe my initial response was simply generated by AI, I'd be very curious to know which major LLM in 2026 will misspell "Boolean" as "Bolean" in `ExpectChildNoBolean`.
